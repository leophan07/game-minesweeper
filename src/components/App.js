import styled from 'styled-components';

export const Wrapper = styled.div`
  text-align: center;
`;

export const Title = styled.h1`
  margin-top: 20px;
`;