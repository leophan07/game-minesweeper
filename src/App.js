import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Wrapper, Title } from './components/App';
import NewGameContainer from './containers/NewGame';
import GridMinesContainer from './containers/GridMines';
import { setData, setDataSuccess, setDataError, setIsPlaying } from './redux/actions';
import { generateData } from './helpers';

function App({ onSetData, onSetDataSuccess, onSetDataError, onSetIsPlaying, isPlaying }) {
  const [mines, setMines] = useState(10);

  async function onPlayGame(level) {
    const mines = level === 9 ? 10 : 40;
    setMines(mines);
    onSetData();
    try {
      const response = await fetch(`https://tiki-minesweeper.herokuapp.com/getMines?size=${level}&mines=${mines}`);
      if (response) {
        const result = await response.json();
        const { data, msg } = result;
        if (msg === 'success') {
          const dataGrid = generateData(level, data);
          onSetDataSuccess(dataGrid);
          onSetIsPlaying();
        }
      }
    } catch (error) {
      onSetDataError(error);
    }  
  }

  return (
    <Wrapper>
      <Title>Mini Minesweeper</Title>
      { !isPlaying ? 
        <NewGameContainer onPlayGame={onPlayGame} /> 
      : 
        <GridMinesContainer onPlayGame={onPlayGame} mines={mines} />
      }
    </Wrapper>
  );
}

App.propTypes = {
  isPlaying: PropTypes.bool.isRequired,
  onSetData: PropTypes.func.isRequired,
  onSetDataSuccess: PropTypes.func.isRequired,
  onSetDataError: PropTypes.func.isRequired,
  onSetIsPlaying: PropTypes.func.isRequired,
}

const mapStateToProps = ({ isPlaying }) => {
  return { isPlaying };
}

const mapDispatchToProps = dispatch => {
  return {
    onSetData: () => {
      return dispatch(setData());
    },
    onSetDataSuccess: data => {
      return dispatch(setDataSuccess(data));
    },
    onSetDataError: err => {
      return dispatch(setDataError(err));
    },
    onSetIsPlaying: () => {
      return dispatch(setIsPlaying());
    },
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
