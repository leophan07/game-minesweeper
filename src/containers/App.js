import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Wrapper, Title } from '../components/App';
import NewGameContainer from './NewGame';
import GridMinesContainer from './GridMines';
import { setData, setDataSuccess, setDataError, setIsPlaying } from '../redux/actions';
import { generateData } from '../helpers';
import { getMines } from '../services/api';

function App({ onSetData, onSetDataSuccess, onSetDataError, onSetIsPlaying, isPlaying }) {
  const [mines, setMines] = useState(10);

  async function onPlayGame(level) {
    const mines = level === 9 ? 10 : 40;
    setMines(mines);
    onSetData();

    try {
      const response = await getMines(level);
      const dataGrid = generateData(level, response);
      onSetDataSuccess(dataGrid);
      onSetIsPlaying();
    } catch (error) {
      onSetDataError(error);
    }
  }

  return (
    <Wrapper>
      <Title>Mini Minesweeper</Title>
      { !isPlaying ? 
        <NewGameContainer onPlayGame={onPlayGame} /> 
      : 
        <GridMinesContainer onPlayGame={onPlayGame} mines={mines} />
      }
    </Wrapper>
  );
}

App.propTypes = {
  isPlaying: PropTypes.bool.isRequired,
  onSetData: PropTypes.func.isRequired,
  onSetDataSuccess: PropTypes.func.isRequired,
  onSetDataError: PropTypes.func.isRequired,
  onSetIsPlaying: PropTypes.func.isRequired,
}

const mapStateToProps = ({ isPlaying }) => {
  return { isPlaying };
}

const mapDispatchToProps = dispatch => {
  return {
    onSetData: () => {
      return dispatch(setData());
    },
    onSetDataSuccess: data => {
      return dispatch(setDataSuccess(data));
    },
    onSetDataError: err => {
      return dispatch(setDataError(err));
    },
    onSetIsPlaying: () => {
      return dispatch(setIsPlaying());
    },
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
