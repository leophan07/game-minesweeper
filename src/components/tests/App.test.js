import React from 'react';
import { render } from '@testing-library/react';

import { Wrapper, Title } from '../App';

describe('<App />', () => {

  it('Wrapper should render an <div> tag', () => {
    const { container } = render(<Wrapper />);
    expect(container.querySelector('div')).toBeTruthy();
  });

  it('Title should render an <h1> tag', () => {
    const { container } = render(<Title />);
    expect(container.querySelector('h1')).toBeTruthy();
  });

  it('App should match the snapshot', () => {
    const { container } = render(
      <Wrapper>
        <Title>Mini MineSweeper</Title>
      </Wrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
