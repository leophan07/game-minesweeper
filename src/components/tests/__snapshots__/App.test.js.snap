// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> App should match the snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa khUSez"
  >
    <h1
      class="sc-bwzfXH bszZAY"
    >
      Mini MineSweeper
    </h1>
  </div>
</div>
`;
